### ISSUE TITLE
Git hygiene

### ISSUE THREAD

**Issue by rasapetter** [rethinkdb/rethinkdb#4537]
With the risk of coming across as an arrogant ass, I've been peeking at the commit history of RethinkDB and find it to be somewhat messy in places. I think some basic guidelines in `CONTRIBUTING.md` and some occasional tough love by the maintainers before merging PRs would be beneficial for all parties. I don't think asking contributors to rebase their branches and clean up their commit messages before a PR is accepted is too much to ask.

For people contributing to the project, guidelines helps reduce confusion about the expectations on building a successful PR. For maintainers, nice commit messages and branches helps with code review. Lastly, for an outsider like me, a pretty commit history helps me build confidence in the maturity and stability of the project as a whole.

Some examples from the last few months:

1) Commit messages without any real content: *"merged"*, *"bugfixes"*, etc.
2) Redundant merges (many instances of *"Merge branch 'next' of github.com:rethinkdb/rethinkdb into next"*)
3) Very long messages (difficult to read in a terminal): *"Changing Tornado python driver Cursor.fetch_next to return True if there is an error on the cursor, so users don't accidentally miss errors - OTS review by @mlucy"*
4) *"WIP: Split ReQL-specific stuff out of B-tree logic."* Why is a work-in-progress in the main branch?
5) *"Ahm the previous commit made all reads stall because I used fifo_enforcer_queue incorrectly."* If the parent commit did something wrong, why not simply remove it?
6) Same with: *"One more abstraction in varint.cc"* right before *"Reverting last changes as it is not directly related to Rethinkdb"*

I haven't contributed to RethinkDB at all and I'm not in position to argue strongly about this. It's only an observation which I felt could have some value if shared. I hope I haven't stepped on any toes, as that was not the intention.

**Comment by danielmewes** [comment]
Hi @rasapetter . Thanks for the great comments.

I think this is less a problem with pull requests rather than with our internal developing process at the company behind RethinkDB.

Keeping the commit log tidy is not a major priority at the moment, since we use GitHub issues together with branches and (internal) code reviews to keep track of when features and fixes are added to the main branch.
We generally try to make every commit at least compile to make `git bisect` more useful, but not all of us (including myself) clean up the commit log before merging a branch. Doing so can also be dangerous sometimes because amending the commit history requires the use of `push --force` which can easily go wrong if the configuration isn't right or one is in the wrong branch.

I'll keep this open for some later discussion, but I don't think we're going to drastically change this in the short term.

**Comment by larkost** [comment]
I would like to argue for always using `--squash` merges when checking into main branches (e.g.: `next` or `v2.1.x`). This would largely eliminate this type of problem since the merge commit would presumably be labeled with the github issue that it was checked in to address. The messy in-flight commit messages then have a lot more context since they are in their own branch. I have been using this method for some time now.

The usual argument against doing this is the one around `bisect`, but as someone who does that fairly often I actually think it improves it in most cases. I generally don't care about the specific checkin that caused the problem, but rather am looking for the reasoning behind why that change was made. So from `next` I would rather be led to the github issue, and the implicit conversation around the issue, then the direct commit for that line of code. If the individual commit is worthwhile, then I can still go into the branch where development was done and `bisect` there.

I also like this workflow because it gives me one last chance to make sure that I have vetted every single line of change relative to where `next` is now. I often get into merge conflicts (nature of the testing system), and having that last clean step to review everything against exactly where it is going really reassures me.

**Comment by mlucy** [comment]
I find it extremely helpful to have intermediate commits when bisecting, especially for large branches.  I think that squashing all commits in branches together when we merge into next would be bad.  (We have some branches that are outstanding for months at a time and build up a lot of changes.)  I wouldn't being opposed to having a rule that all commits need to compile, and squashing down commits that don't.  (I usually try to do this in my own branches but I sometimes miss some.)

I think long commit messages and commits like "merged" are fine.  Commits like "bugfixes" should probably be merged into the commit before them.

**Comment by rasapetter** [comment]
@mlucy Let me rephrase my point about long commit messages. Long messages are great, the more descriptive the better. However, since people tend to read commits in places where overflowing lines aren't handled properly (terminal, gitx and even github), it's good practice to break the lines manually.

**Comment by deontologician** [comment]
As far as commit messages, there is a standard that's encouraged pretty
much everywhere that the first line be < 50 characters, then an empty line,
then the longer in-depth description. I don't see any reason not to follow
that.

**Comment by danielmewes** [comment]
I like the idea of using `merge --squash`. That should make the commit history on the main branch much easier to follow.
Let's do that!

Making the first line short and then having details after an empty line also sounds good. We should try to do that when possible.

(mentioning the relevant people who haven't commented here yet so everyone is aware of this: @Tryneus, @VeXocide, @AtnNn)

**Comment by mlucy** [comment]
There are definitely times when squashing an entire branch down is inappropriate (the Raft branch, for example).  What if we instead just had a rule that every commit had to compile, and people could use their judgement as to whether to squash the branch into `next` or clean up the history before doing a normal merge?

**Comment by danielmewes** [comment]
Good point @mlucy. I didn't mean to set an absolute rule about this.
Certainly this shouldn't apply to long-running branches like `raft` (though it probably should have applied to the sub-branches that were merged into `raft` before `raft` got merged into `next`).

Revised non-rule: Please use `git merge --squash` where reasonable (usually small- to medium-size changes), and otherwise perform reasonable cleanup in the branch itself, followed by a regular merge.

**Comment by deontologician** [comment]
Just a note, if you're rebasing a lot of commits and just want to squash some of them:

```
git rebase -i `git merge-base feature_branch next`
```
